### ISSUE TITLE
OpenTele measurements fail w/ status code 404

### ISSUE THREAD

**Issue by rtud** [cami-project/cami-project#140]
## Why
For the CAMI consortium, it is extremely important for the stakeholders to have the systems they have built integrated.

Following the checks in #136 to insure measurements are being sent to the OpenTele API accordingly, it seems that the integration is no longer working properly and posting the measurements over returns w/ a HTTP status code of 404. 

## What
* [x] the cause for the measurements failing to update is investigated and a resolution is provided as a comment on this issue

## Notes
Here is the current log message when attempting to post to the OpenTele API:
```
Feb 27 11:35:48 139.59.181.210 logger:  <15>DEBUG 2017-02-27 09:35:48,002 endpoints 19 140227251435264 [opentele] The result of posting {'user_id': 11262861, 'timestamp': 1488195123, 'value': 69.0, 'measurement_unit': 'bpm', 'input_source': 'test', 'timezone': 'Europe/Bucharest', 'type': 'heartrate'} to OpenTele: 404
```

**Comment by rtud** [comment]
@aismail & @asorici: After a thorough investigation into the 404 errors, I have found that most likely there are some issues w/ the OpenTele API. Following are some simple cURL commands that reproduce the issues:

### Pulse
Having run multiple tests, using Postman and cURL, I've found that we can only successfully post a Pulse value larger than `90`. Any lower, and we get a 404 response:

**PULSE = 90**
```bash
curl -v -H "Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=" --user nancyann:abcd1234 -d '{"name": "CAMIBloodPressure", "date": "2017-02-28T08:30:00.000+02:00", "QuestionnaireId": 53, "output": [{"name": "332.BP#PULSE", "type": "Int", "value": 90}, {"name": "332.BP#SEVERITY", "type": "String", "value": "GREEN"}]}' http://opentele.aliviate.dk:4288/opentele-citizen-server/rest/questionnaire/listing/
*   Trying 5.103.116.59...
* TCP_NODELAY set
* Connected to opentele.aliviate.dk (5.103.116.59) port 4288 (#0)
* Server auth using Basic with user 'nancyann'
> POST /opentele-citizen-server/rest/questionnaire/listing/ HTTP/1.1
> Host: opentele.aliviate.dk:4288
> Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> Content-Length: 224
>
* upload completely sent off: 224 out of 224 bytes
< HTTP/1.1 404 Not Found
< Server: Apache-Coyote/1.1
< X-MiniProfiler-Ids: ["fc4f1ee3-7d3e-40eb-a1d0-329fc2a1b62c"]
< Set-Cookie: JSESSIONID=A41365DFA2D55B0456575E2ED360B9DF; Path=/opentele-citizen-server; HttpOnly
< Content-Type: text/html;charset=ISO-8859-1
< Content-Language: en
< Content-Length: 1085
< Date: Tue, 28 Feb 2017 09:10:46 GMT
<
* Curl_http_done: called premature == 0
* Connection #0 to host opentele.aliviate.dk left intact
<html><head><title>Apache Tomcat/7.0.72 - Error report</title><style><!--H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}HR {color : #525D76;}--></style> </head><body><h1>HTTP Status 404 - &quot;/WEB-INF/grails-app/views/productionError.gsp&quot; not found.</h1><HR size="1" noshade="noshade"><p><b>type</b> Status report</p><p><b>message</b> <u>&quot;/WEB-INF/grails-app/views/productionError.gsp&quot; not found.</u></p><p><b>description</b> <u>The requested resource is not available.</u></p><HR size="1" noshade="noshade"><h3>Apache Tomcat/7.0.72</h3></body></html>
```

**PULSE = 91**
```bash
curl -v -H "Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=" --user nancyann:abcd1234 -d '{"name": "CAMIBloodPressure", "date": "2017-02-28T08:30:00.000+02:00", "QuestionnaireId": 53, "output": [{"name": "332.BP#PULSE", "type": "Int", "value": 91}, {"name": "332.BP#SEVERITY", "type": "String", "value": "GREEN"}]}' http://opentele.aliviate.dk:4288/opentele-citizen-server/rest/questionnaire/listing/
*   Trying 5.103.116.59...
* TCP_NODELAY set
* Connected to opentele.aliviate.dk (5.103.116.59) port 4288 (#0)
* Server auth using Basic with user 'nancyann'
> POST /opentele-citizen-server/rest/questionnaire/listing/ HTTP/1.1
> Host: opentele.aliviate.dk:4288
> Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> Content-Length: 224
>
* upload completely sent off: 224 out of 224 bytes
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< X-MiniProfiler-Ids: ["5ab055ec-fc53-45e6-9dde-6b32a6ad8121"]
< Set-Cookie: JSESSIONID=251CBAF0716DA7A6115AF24980C6ADE1; Path=/opentele-citizen-server; HttpOnly
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Date: Tue, 28 Feb 2017 09:10:14 GMT
<
* Curl_http_done: called premature == 0
* Connection #0 to host opentele.aliviate.dk left intact
[["success"],["Received 2 variables. Stored 1 questionnaire(s), 1 result(s) and 1 measurement(s)."]]
```

### Weight
We are currently unable to post the weight measurements. I've tested this with the available Postman setup, as well as with a basic cURL command:

```bash
curl -v -H "Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=" --user nancyann:abcd1234 -d '{"name": "CAMIWeight", "date": "2017-02-28T08:30:00.000+02:00", "QuestionnaireId": 55, "output": [{"name": "332.WEIGHT", "type": "Int", "value": 77}, {"name": "336.WEIGHT#SEVERITY", "type": "String", "value": "GREEN"}]}' http://opentele.aliviate.dk:4288/opentele-citizen-server/rest/questionnaire/listing/
*   Trying 5.103.116.59...
* TCP_NODELAY set
* Connected to opentele.aliviate.dk (5.103.116.59) port 4288 (#0)
* Server auth using Basic with user 'nancyann'
> POST /opentele-citizen-server/rest/questionnaire/listing/ HTTP/1.1
> Host: opentele.aliviate.dk:4288
> Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: application/json, Authorization: Basic bmFuY3lhbm46YWJjZDEyMzQ=
> Content-Length: 219
>
* upload completely sent off: 219 out of 219 bytes
< HTTP/1.1 404 Not Found
< Server: Apache-Coyote/1.1
< X-MiniProfiler-Ids: ["1ed31b20-bfc3-4db8-b35f-384d356d84b5"]
< Set-Cookie: JSESSIONID=BB92AC2E1210FCF8229EFFC82736C954; Path=/opentele-citizen-server; HttpOnly
< Content-Type: text/html;charset=ISO-8859-1
< Content-Language: en
< Content-Length: 1085
< Date: Tue, 28 Feb 2017 09:09:27 GMT
<
* Curl_http_done: called premature == 0
* Connection #0 to host opentele.aliviate.dk left intact
<html><head><title>Apache Tomcat/7.0.72 - Error report</title><style><!--H1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} H2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} H3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} BODY {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} B {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} P {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;}A {color : black;}A.name {color : black;}HR {color : #525D76;}--></style> </head><body><h1>HTTP Status 404 - &quot;/WEB-INF/grails-app/views/productionError.gsp&quot; not found.</h1><HR size="1" noshade="noshade"><p><b>type</b> Status report</p><p><b>message</b> <u>&quot;/WEB-INF/grails-app/views/productionError.gsp&quot; not found.</u></p><p><b>description</b> <u>The requested resource is not available.</u></p><HR size="1" noshade="noshade"><h3>Apache Tomcat/7.0.72</h3></body></html>
```

Please let me know if I can provide you with additional information on this issue, thank you!

**Comment by asorici** [comment]
@aismail and @rtud 
I've gotten reply from Aliviate on how to fix the error we're having. Reason for the errors is database corruption on the http://opentele.aliviate.dk:4288/opentele-citizen-server/ machine when sending values with severity level GREEN.
Debugging why this happens is beyond our scope.

Fix:
  - change URI of OpenTele endpoint from http://opentele.aliviate.dk:4288/opentele-citizen-server/ to
    http://opentele.aliviate.dk:4388/opentele-citizen-server/ (notice the change in port number - 4388)

  - use the following values for QuestionnaireId:
    - CAMIWeight: 51
    - CAMISaturation: 52
    - CAMIBloodPressure: 50

  - use the following key prefix for BP body: 324.BP
  
  - use the following key prefix for WEIGHT body: awaiting reply to email

**Comment by asorici** [comment]
key prefix for WEIGHT body: 326.WEIGHT

**Comment by rtud** [comment]
Thanks for the updates @asorici, I'll take care of the changes no later than early next week.

**Comment by aliviateeh** [comment]
Keyprefix:

Weight: 
326.WEIGHT

Saturation: 
328.SAT#SATURATION
328.SAT#PULSE

Please consider dynamic implementation instead if you got the time ...

**Comment by rtud** [comment]
This issue has been resolved thanks to @asorici and we've successfully tested the correct functionality in #176 & #178. The reason why we weren't seeing the updated measurements was described by @cretualexandru [here](https://github.com/cami-project/cami-project/issues/176#issuecomment-289010826).

@asorici also pointed another case when we get 404 errors, when attempting to post 2 measurements in rapid succession. This will be tended to in #185.

Closing, thank you!
